WBooks/AddNew/AddNewController.swift

```diff
@@ -71,42 +71,56 @@ class AddNewController: UIViewController {
     }
     
     private func setSubmitButton() {
+        setSubmitEnableLogic()
+        setSubmitAction()
+    }
+    
+    private func setSubmitEnableLogic() {
         _formFilled.producer.startWithValues { [weak self] formFilled in
-            if let this = self {
-                if formFilled && !this._loading {
-                    this._view.enableSubmit()
-                } else {
-                    this._view.disableSubmit()
-                }
+            guard let this = self else {
+                return
+            }
+            if formFilled && !this._loading {
+                this._view.enableSubmit()
+            } else {
+                this._view.disableSubmit()
             }
         }
-        _view.submitButton.reactive.controlEvents(.touchUpInside).observeValues { [weak self] _ in
-            self?._loading = true
-            self?._view.disableSubmit()
-            self?._view.disableInteractions()
-            self?.submit()
-        }
-    }
-    
-    func submit() {
-        _viewModel.addBook(onSuccess: onAddNewSuccess, onError: onError)
     }
     
-    func onAddNewSuccess() {
-        _loading = false
-        _view.resetForm()
-        _view.enableInteractions()
-        let alert = UIAlertController(alertViewModel: ErrorAlertViewModel(title: "THANKS".localized(), message: "BOOK_ADDED".localized(), dismissButtonTitle: "ACCEPT".localized()))
-        self.present(alert, animated: true, completion: nil)
-    }
-    
-    func onError(error: Error) {
-        _loading = false
-        _view.enableSubmit()
-        _view.enableInteractions()
-        let alert = UIAlertController(alertViewModel: ErrorAlertViewModel(title: "UPS".localized(), message: "BOOK_ADD_ERROR".localized(), dismissButtonTitle: "ACCEPT".localized()))
-        self.present(alert, animated: true, completion: nil)
-        print(error)
+    private func setSubmitAction() {
+        //Submit button action
+        _view.submitButton.reactive.controlEvents(.touchUpInside).observeValues { [weak self] _ in
+            guard let this = self else {
+                return
+            }
+            this._loading = true
+            this._view.disableSubmit()
+            this._view.disableInteractions()
+            this._viewModel.addBook()
+        }
+        //Actions when submit request is completed
+        _viewModel.addState.producer.startWithValues { [weak self] state in
```

Comment:
You can use reactive swift `Actions` to manage this kind of requests. With that you can observe values and errors and manage each separately without the enum. You don't need to change it in this training, but keep it in mind :D

Reply:
Cool =D